Allow customisation of fields logged in middleware

ianyong opened the pull request:
## Context

Currently, it is not possible to specify additional fields for the `SuperloggerMiddleware` to log. This means that consumers that wish to do so need to monkey patch Superlogger.

## Changes

Add a new configuration option `log_extra_fields` which can be optionally defined as a method that takes in an `ActionDispatch::Request` object and returns a hash of key-value pairs to log. This change is backward-compatible.

## How to Verify

See the new test case added.

ianyong committed:
Allow customisation of fields logged in middleware

ianyong committed:
Merge branch 'master' into allow-customisation-of-fields-logged-in-middleware

ianyong committed:
Add test case for logging extra fields

ianyong committed:
Add instructions on logging extra fields to README

ianyong commented on Gemfile.lock:
```diff
@@ -1,7 +1,7 @@
   remote: .
   specs:
-    superlogger (1.0.1)
+    superlogger (1.0.2)
```
Apparently building the gem also bumps this 🙈 Will take note of this for the next release.

dorcastan replied:
Suggestion: How about adding release instructions to the `README`? (Doesn't have to be part of this PR)

ianyong replied:
The proper solution would honestly be a CD job because there are other aspects. I looked inside the gemfile and noticed that all of the temporary files from running the test cases were also bundled inside, which is why the gem increased in size from the previous version - a CD job would resolve this.

ianyong replied:
I'll do that as a stopgap measure since it doesn't make too much sense to put too much effort here, thanks!

ianyong replied:
Added publish instructions to README in https://github.com/moexmen/superlogger/pull/26.

dorcastan replied:
Thank you!

> I looked inside the gemfile and noticed that all of the temporary files from running the test cases were also bundled inside, which is why the gem increased in size from the previous version - a CD job would resolve this.

🙈

dorcastan commented on test/superlogger_test.rb:
```diff
@@ -64,6 +64,41 @@ def output
+  # The extra fields to log are defined in `/test/dummy/config/application.rb`.
+  # Unfortunately, because `Superlogger::SuperloggerMiddleware` is initialised
+  # together with the Rails application, and because we cannot programmatically
+  # reinitialise the Rails application or run multiple instances of it, we are
+  # only able to test a single Superlogger configuration. Once the Rails
+  # application is initialised, the middleware stack cannot be changed.
```
Suggestion, to simplify slightly / improve readability.

```suggestion
  # reinitialise the Rails application or run multiple instances of it, we can
  # only test a single Superlogger configuration. Once the Rails application is
  # initialised, the middleware stack cannot be changed.
```

dorcastan commented on README.md:
````diff
@@ -73,6 +73,17 @@ Rails.logger.info "Meatball"
 
+To log additional fields that are available on the `ActionDispatch::Request` object, set in `config/application.rb`:
+```ruby
+Superlogger.log_extra_fields = lambda do |request|
````
I'm wondering if there's a better name for this property.

1. "Extra" seems to have some connotation of "unnecessary/not strictly needed", which feels a little odd. 😅 
2. This method doesn't do the actual logging (which might be implied by `log` as a verb), but more configures which fields should be logged.

Ideas:
* `include_fields`
* `include_log_fields`
* `other_log_fields`
* `additional_log_fields` (this was one of the first that came to mind, but it's long)

This is non-blocking if we can't find a better name, though.

ianyong replied:
I'm in favour of `include_log_fields` for the reasons you've stated - will make the change, thanks!

dorcastan commented on test/superlogger_test.rb:
```diff
@@ -64,6 +64,41 @@ def output
+  # only able to test a single Superlogger configuration. Once the Rails
+  # application is initialised, the middleware stack cannot be changed.
+  #
+  # This effectively means that we are unable to change the Superlogger settings
+  # in the Rails application configuration on a per-test basis. For the sake of
+  # testing, we choose the Superlogger settings that let us test the presence of
+  # behaviours and set the extra fields to log.
```
For documentation: To frame this differently, this means that we cannot test that the logging middleware works even when `Superlogger.log_extra_fields` is undefined (nil)?

ianyong replied:
No, we can test that the logging middleware works when `Superlogger.log_extra_fields` is `nil`, just not in the same test suite run as another test case which requires that `Superlogger.log_extra_fields` is set to some other value.

Specifically, Superlogger is set to initialise itself together with Rails:
https://github.com/moexmen/superlogger/blob/97d3370885e6f5bdccfe7e93bc4607d50dec61bb/lib/superlogger/railtie.rb#L1-L7

After Superlogger's initialisation, any change made to its settings doesn't do anything as the `setup` method is only called once:
https://github.com/moexmen/superlogger/blob/97d3370885e6f5bdccfe7e93bc4607d50dec61bb/lib/superlogger.rb#L9-L15

Calling the `setup` method manually doesn't work because Rails freezes its middleware stack after initialisation. Additionally, there is no way to programmatically unload Rails to load it again with a different configuration.

It might be possible to use a gem such as [`combustion`](https://github.com/pat/combustion) to workaround the above limitations, but it will likely require overhauling the entire testing setup (including modernising it to use RSpec).

ianyong committed:
Rename `log_extra_fields` to `include_log_fields`

For the following reasons raised by Dorcas:
1) "Extra" seems to have the connotation of "unnececssary/not
   strictly needed" which feels a little odd.
2) The method doesn't do the actual logging (which might be implied
   by the use of `log` as a verb), but more configures which fields
   should be additionally logged.

dorcastan reviewed (approved):
Thanks!
